---
layout: page
title: Basic Bioconductor infrastructure
---

```{r options, echo=FALSE}
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```
# Basic Bioconductor infrastructure

```{r}
source("http://bioconductor.org/biocLite.R")
installifnot <- function (packageName){
  if (!(require(packageName, character.only=TRUE))) biocLite(packageName)
}
installifnot("IRanges")
installifnot("GenomicRanges")
installifnot("GenomicAlignments")
installifnot("parathyroidSE")
installifnot("GenomicFeatures")
```

## IRanges

IRanges is a library for representing ranges of integers, which is useful in genomics, because we have base pair
ranges that we'd like to manipulate. There's a very detailed vignette, which you can access, which contains a lot of detailed information about all the different classes and functions which are contained in the `IRanges`
Package.

This lab shows a summary of the most important functions.

IRanges have a start an end, and a width.
If you just specify two numbers, you specify the start and the end.


```{r}
library(IRanges)
ir <- IRanges(5,10)
ir
start(ir)
end(ir)
width(ir)
# ?IRanges
```

Here we have an IRange which starts at 5, ends at 10,
and it's there for six base pairs long.
You can see that the first three
arguments are start, end, and width.
And if you specify the start is 5, and the width should be 6,
you'll get the identical IRange.

```{r}
ir <- IRanges(start=c(3,5,17), end=c(10,8,20))
ir
ir <- IRanges(5,10)
```

There are a number of intra range methods for IRanges.
And intra range means that the operation will occur for each range
that you have, and it doesn't depend on the other ranges that
are contained in IRanges object.

An example of this is to shift the I Range to the left by two.
Before we had an I Range that started at 5 and ended at 10,
and applying the shift operation produces an I Range which starts at 3
and ends at 8.
```{r}
# ?"intra-range-methods"
shift(ir, -2)
```

Remeber, all of these commands can work on more than one range at once. Here we show the effects of the different methods using a single range:

In addition to the shift function, we also have a number of other functions,
including `narrow`. THis function works relative to the start, you should instead
start this range at the second base pair.
So relative to 5, it should start at 6.
You can also say, relative to 5, you should
end on the fifth base pair, which means it should end at 9.

The `flank` function allows you to get flanking sequence
here three base pairs from the start.
You can also get three base pairs from the end
by specifying start equals false.
And you can say you want a bidirectional flanking sequence
from the start by specifying both equals true.

```{r}
ir
shift(ir,-2)
narrow(ir, start=2)
narrow(ir, end=5)
flank(ir, width=3, start=TRUE, both=FALSE)
flank(ir, width=3, start=FALSE, both=FALSE)
flank(ir, width=3, start=TRUE, both=TRUE)
ir * 2
ir + 2
ir - 2
```

The previous functions and operations can be easily visualized simply using the `plot` function.

```{r}
# set up a plotting window so we can look at range operations
plotir <- function(ir,i) { arrows(start(ir)-.5,i,end(ir)+.5,i,code=3,angle=90,lwd=3) }
plot(0,0,xlim=c(0,15),ylim=c(0,11),type="n",xlab="",ylab="",xaxt="n")
axis(1,0:15)
abline(v=0:30 + .5,col=rgb(0,0,0,.5))

# plot the original IRange
plotir(ir,1)

# draw a red shadow for the original IRange
polygon(c(start(ir)-.5,start(ir)-.5,end(ir)+.5,end(ir)+.5),c(-1,12,12,-1),col=rgb(1,0,0,.2),border=NA)
plotir(shift(ir,-2), 2)
plotir(narrow(ir, start=2), 3)
plotir(narrow(ir, end=5), 4)
plotir(flank(ir, width=3, start=TRUE, both=FALSE), 5)
plotir(flank(ir, width=3, start=FALSE, both=FALSE), 6)
plotir(flank(ir, width=3, start=TRUE, both=TRUE), 7)
plotir(ir * 2, 8)
plotir(ir + 2, 9)
plotir(ir - 2, 10)
```

The inter range methods are those functions which depend on the other ranges in the object.
We can create an IRanges object with three ranges, which
starts at 3, 5, 17, ends at 10, 8, and 20. 
If you ask for the range, this will give you the beginning of the IRanges
to the end, including gaps in between.
So this gives us the start of 3 and the end of 20.

If we say `reduce`, it gives us those base pairs
which are covered by the original ranges.
So we do not get the gap, so the end at 10 and the beginning at 17.

We can ask for the gaps, which gives us a gap from 10 to 17.
Therefore, the gap starts at 11 and ends at 16.

Finally, we have the `disjoint` function, which
gives a set of ranges which has the same coverage as the original IRanges
object, but they're not overlapping in any way,
and they also contain the union of all the endpoints of the original range.

```{r}
# ?"inter-range-methods"
ir <- IRanges(start=c(3,5,17), end=c(10,8,20))
range(ir)
reduce(ir)
gaps(ir)
disjoin(ir)
```

## GRanges and GRangesList

Next we'll introduce the `GenomicRanges` package,
which is an extension of `IRanges` to the genomic space.
Most importantly these contain a sequence name.

### GRanges

We see below an IRange on chromosome z,
and it can also contain strand information and sequence lengths.

Sequences correspond generally to chromosomes.
So here if we print out the GRanges object,
we see that we have two ranges, zero metadata columns.

It gives the sequence names as an rle, a "run length encoding".
It gives the IRanges and the strand, also as an rle 

```{r}
library(GenomicRanges)
gr <- GRanges("chrZ", IRanges(start=c(5,10),end=c(35,45)),
              strand="+", seqlengths=c(chrZ=100L))
gr
shift(gr, 10)
shift(gr, 80)
trim(shift(gr, 80))
mcols(gr)
mcols(gr)$value <- c(-1,4)
gr
```

### Rle or "run-length encodings"

This is an object which is defined by IRanges.
But also there's a similar object in base r which is a run length encoding.
And the meaning of this is that if you have
a vector, which repeats certain values, you can save memory.
By instead of storing each value, you save the number, and then
the number of times that that number's repeated.

Let's see an example.

```{r}
r <- Rle(c(1,1,1,0,0,-2,-2,-2,rep(-1,20)))
r
str(r)
as.numeric(r)
```

Here we have a vector with certain repeated values.
For instance, the last one is repeated 20 times.

If we save this as an rle, using the rle function,
we see that it's represented, in a compact form,
with the values and the lengths that each value is repeated.
The first value, 1, is repeated three times.
And then we get a repeat of zero.
And then value -2 is repeated three times.

If we use the str function, which gives our compact presentation,
we see that in fact it's just storing these values.
Instead of storing all the numbers, we're
just storing the compact representation.

If we have such an rle object, and we want
to peer into it in different regions, we can construct a views object.
Views is a virtual class, which contains the subject,
and then a number of views, which are essentially IRanges into that object.
So for example, we had this rle R, and let's say we want to create two views.
One which starts at 4 and ends at 7, a second which starts at 2 and ends at 6.
If we enter this in, we can see we've constructed
this view, which starts at 4 and 7.
And then it peers into the rle, and gives us
those values which were contained there.


```{r}
Views(r, start=c(4,2), end=c(7,6))
```


It's actually only a virtual class, in that
it didn't deconstruct the first rle, but instead we've just saved the rle and a number of windows or views into it.

This is not only for use with rle.
You can also use the views constructor for FASTA files,
for example, if you want to look into genome sequence or other objects.

### GRangesList

GRangesList is an object which groups GRanges together.

```{r}
gr2 <- GRanges("chrZ",IRanges(11:13,51:53))
mcols(gr)$value <- NULL
grl <- GRangesList(gr,gr2)
grl
length(grl)
grl[[1]]
mcols(grl)$value <- c(5,7)
grl
mcols(grl)
```

The most obvious example of a GRangesList would be grouping exons by gene,
or grouping axons by transcript.

You can ask for the length and this is giving the number of GRanges.
And you can ask for the first element, which returns the first GRanges.
And if you specify metadata columns to the GRangesList,
these will be assigned to each GRanges object in the list.

### findOverlaps and %over%

Once we've created sets of GRanges or GRanges list objects,
one common thing we might need to do is to find overlaps between objects.

```{r}
gr1 <- GRanges("chrZ",IRanges(c(1,11,21,31,41),width=5))
gr2 <- GRanges("chrZ",IRanges(c(19,33),c(38,35)))
gr1
gr2

```

It looks like they might have some overlap.
So we'll use the `findOverlaps` function to find the overlaps.

If you look at that help page for `findOverlaps`,
you can see that there's a lot of different arguments here.
But the first two, `query` and `subject`, are the most important.

```{r}
fo <- findOverlaps(gr1, gr2)
fo
queryHits(fo)
subjectHits(fo)
```

The result of the `findOverlaps` function is `hits` object, which has length three,
and this gives us the three different overlaps which occurred.

The table here tells us that the third element of the query
intersected with the first element of the subject.
And these are given as integer vectors.
The third element of gr1, intersected with the first element of gr2.
And likewise the fourth intersected with the first, and the fourth intersected
with the second.
We can see that indeed the fourth element of gr1
does overlap the second element of gr2.


Another way to get after this is to use the over function, `%over%`, which gives a logical vector.
So for the five ranges in gr1, it gives a logical vector
describing which of these have any overlap with the ranges in the second.

```{r}
gr1 %over% gr2
gr1[gr1 %over% gr2]
```



## SummarizedExperiment

SummarizedExperiments, similarly to `Rle` or the (in)famous `expressionSet` class of microarray data are classes intended to store complex data structures

### Summarized Experiment

We use a bioconductor annotation package, which
is the `parathyroid SummarizedExperiment library`. 
It contains a `SummarizedExperiment` object, which summarizes counts of RNA sequencing
reads in genes and ensemble genes for an experiment on human cell culture.

```{r}
library(parathyroidSE)
data(parathyroidGenesSE)
se <- parathyroidGenesSE
se
```

We can see that we have a SummarizedExperiment object with 63,000
rows, which are genes, and 27 columns, which
are samples, and the matrix, in this case, is called `counts`.

We also have:
- the row names, which are ensemble genes, 
- metadata about the row data, and 
- metadata about the column data.

We can access the counts matrix, which contains
the counts of RNA sequencing reads, using the `assay` function.

```{r}
dim(se)
assay(se)[1:3,1:3]
dim(assay(se))
```

The `colData` function gives access to the covariates information. For those familiar with microarrays it is the equivalent to `pData` function in the `ExpressionSet` class.

```{r}
colData(se)[1:3,1:6]
dim(colData(se))
names(colData(se))
colData(se)$treatment
```

The `rowData` function gives access to the rows, which,  in this case, correspond to genes, that is collections of exons.

The rows of the SummarizedExperiment is a GRangesList
where each row corresponds to a GRanges which contains the exons, which
were used to count the RNA sequencing reads.

```{r}
rowData(se)[1]
class(rowData(se))
length(rowData(se))
head(rownames(se))
```

There's also included in this class some metadata,
which is accessible with the `metadata` function.

This is information which is stored which
is telling us about how this `GRangesList` was constructed.

- It was constructed from the genomic features package using a transcript database.
- It's a human, Homo sapiens was the organism, and 
- the database was ENSEMBL GENES number 72, and 
- etc.

So that's a nice way to allow others to know exactly how this thing was constructed.

In addition, there's some more information under experiment data,
using `exptData` and then specifying the `MIAME`, which
is *minimal information about a microarray experiment* (although we're not using microarrays anymore , we've still used the same slots to describe extra information about this object).

```{r}
metadata(rowData(se))
exptData(se)$MIAME
abstract(exptData(se)$MIAME)
```

## Footnotes

For more information about the `GenomicRanges` package, check out the PLOS Comp Bio paper, which the authors of GenomicRanges published:

<http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118>

Also the software vignettes have a lot of details about the functionality. Check out "An Introduction to Genomic Ranges Classes":

<http://www.bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.pdf>

All of the vignette PDFs are available here:

<http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html>

